RecCalorimeter/src/components/CorrectCaloClusters.cpp

```diff
@@ -318,6 +377,118 @@ StatusCode CorrectCaloClusters::applyDownstreamCorr(const edm4hep::ClusterCollec
   return StatusCode::SUCCESS;
 }
 
+StatusCode CorrectCaloClusters::applyBenchmarkCorr(const edm4hep::ClusterCollection* inClusters,
+                                                    edm4hep::ClusterCollection* outClusters) {
+
+  const size_t numReadoutNames = m_readoutNames.size();
+
+  int ecal_index = -1;
+  int hcal_index = -1;
+
+  // Identify ECal and HCal readout positions in the input parameters when running the calibration
+  for (size_t i=0; i<numReadoutNames; ++i){
+    if (m_systemIDs[i] == m_systemIDECal){
+      ecal_index = i; 
+    }
+    else if (m_systemIDs[i] == m_systemIDHCal){
+      hcal_index = i; 
+    }
+  }
+
+  for (size_t j = 0; j < inClusters->size(); ++j) {
+    double energyInLastLayerECal = getEnergyInLayer(inClusters->at(j),
+                                                    m_readoutNames[ecal_index],
+                                                    m_systemIDs[ecal_index],
+                                                    m_lastLayerIDs[ecal_index]);
+  
+    double energyInFirstLayerECal = getEnergyInLayer(inClusters->at(j),
+                                                     m_readoutNames[ecal_index],
+                                                     m_systemIDs[ecal_index],
+                                                     m_firstLayerIDs[ecal_index]);
+
+    double energyInFirstLayerHCal = getEnergyInLayer(inClusters->at(j),
+                                                     m_readoutNames[hcal_index],
+                                                     m_systemIDs[hcal_index],
+                                                     m_firstLayerIDs[hcal_index]);
+
+    double totalEnergyInECal = getTotalEnergy(inClusters->at(j),
+                                         m_readoutNames[ecal_index],
+                                         m_systemIDs[ecal_index]);
+
+    double totalEnergyInHCal = getTotalEnergy(inClusters->at(j),
+                                         m_readoutNames[hcal_index],
+                                         m_systemIDs[hcal_index]);
+                                         
+
+    // calculate approximate benchmark energy using non energy dependent benchmark parameters
+    double approximateBenchmarkEnergy = 0.0; 
+    approximateBenchmarkEnergy = m_benchmarkParamsApprox[0] * totalEnergyInECal +
+                                 m_benchmarkParamsApprox[1] * totalEnergyInHCal +
+                                 m_benchmarkParamsApprox[2] * sqrt(abs(energyInLastLayerECal * m_benchmarkParamsApprox[0] * energyInFirstLayerHCal * m_benchmarkParamsApprox[1])) +
+                                 m_benchmarkParamsApprox[3] * pow(totalEnergyInECal * m_benchmarkParamsApprox[0], 2) +
+                                 m_benchmarkParamsApprox[4] * energyInFirstLayerECal +
+                                 m_benchmarkParamsApprox[5];
+
+    // Calculate energy-dependent benchmark parameters p[0]-p[5]
+    int benchmarkFormulaIndexHighEne = 0;
```

Comment:
Can you instead of declaring an index create new vector out of only high energy formulas?
```
auto benchmarkFormulaHighEne = m_benchmarkFunctions.at(0);
```

Reply:
it should be OK now with more safety checks